### ISSUE TITLE
case-sensitive sorting of results and DataTables warning: table id=d-tables - Ajax error.

### ISSUE THREAD

**Issue by gibsonjc** [CARLI/web-reports#93]
7/22 - I just noticed that the sorting of the results tables may be taking upper case vs. lower case into account.  When I sort these results by “Item Permanent Location”, the location “Dean” comes before “archives”.  Is that a setting that we might be able to adjust to ignore case when sorting alphabetically?  Example: QCYdb on devel, "Item Records with Perm Loc Different from MFHD Loc"

7/25 - Nice catch.  This looks to be fixed.  Should I push the change to prod?

7/25 - Sure, that looks fixed.

8/1 - Noticed with some other unrelated reports ( #52 #92 ) that trying to resort the columns caused error: "DataTables warning: table id=d-tables - Ajax error. For more information about this error, please see http://datatables.net/tn/7"

8/2 1:45pm - This problem is related to the fix pushed to prod last week. The problem is my 'fix' failed to account for the fact that it will at some point be sorting non-string data (this was a pretty dumb mistake for me to make). I'm working on fixing it now. The point of this update is just to make you aware that the problem you're reporting in this particular issue is present now in production and will be triggered when a user attempts to sort a column in any report that contains non-string data in that column (dates & integers pretty much), and it will also automatically manifest itself when the report's first column contains non-string data on table load (because the system initially sorts the first column before display). ... Pulled the offending change from production until I can figure out a more appropriate way to sort columns (with strings!) and ignore the case of the string. Columns will sort without error once again, they've just reverted to their old behavior.

8/2 5:29pm - I believe I have fixed the problem and pushed the change to production.

**Comment by gibsonjc** [comment]
I made an Issue for this one since it has all these side effects.

I confirm that case-insensitive sorting appears to be working correctly (again).
Example: QCYdb on devel, "Item Records with Perm Loc Different from MFHD Loc"

However, now I am getting the error "DataTables warning: table id=d-tables - Ajax error. For more information about this error, please see http://datatables.net/tn/7" when I click on report Results with zero hits.  This is an improvement since now I can see the actual Results with more than zero in HTML, but feels very untidy.

Examples returning the error with zero hits in KNXdb on devel:
Bib record 035 $a has malformed OCoLC prefix - 0
Bib record 035 $a begins ocmocm- 0
Bib record 035 $a begins OCoLC but has invalid characters - 0 
Bib record 035 $a begins OCoLCocm but has more than 8 digits - 0
Bib record 035 $a begins OCoLCocm but lacks leading zeroes - 0

**Comment by gibsonjc** [comment]
With Report Circulation Stat 7: Charges by Circ Hap Loc Item Type and Item Loc:
The "Charges" column now sorts correctly when before it gave the datatables error.  I'm still getting the error on DPUdb on devel, dates Jan. 1-15, 2016: error when I try to sort on the "Item type" table. All other columns re-sort just fine. I note there are a few entries with blank "Item type" lines.  I haven't been able to get this query to complete in Access so I'm not sure what those blanks have for data in its output, but I'll keep trying.  It might be data replacement like Chris did with the other columns in #52 .

**Comment by patrickzurek** [comment]
That's unfortunate.  I've pulled the 'fix' from production for the time being while I work on devel.

**Comment by patrickzurek** [comment]
I think I've implemented a working fix for this.  Both for the issue in @gibsonjc last comment about "Report Circulation Stat 7: Charges by Circ Hap Loc Item Type and Item Loc" when sorting by the "Item type" column.  This should now allow sorting for all queries that contain empty data in some of their column fields.  And it should fix the issue reported in the comment just above it for queries returning zero hits but still throwing a datatable error messages.  Please test it out on devel.  BTW I checked it is indeed ignoring case when sorting alphabetically.  See _iPod_ here in 3rd column (Item Records with Perm Loc Different from the MFHD Loc - QCYDB):  

<img width="1061" alt="screen shot 2016-08-09 at 5 02 42 pm" src="https://cloud.githubusercontent.com/assets/6615138/17535264/262f175e-5e53-11e6-8622-f6937657ed16.png">

Also, I would like to test how the fix performs when presented with data to be sorted that's not represented by the ASCII subset of characters.  I ran a query that returned a large amount of book titles where I had hoped to find an instance of an obvious non-ascii character, but I couldn't find any.  Does anyone know of any reports + input that would return results with non-ascii characters in it?

**Comment by gibsonjc** [comment]
This is a big improvement!  I retested all of the results that I had problems sorting and they are working correctly now, including Circ Stat 7 and the reports that returned zero hits.  I no longer get the datatables error when viewing any results.

Are you looking for examples like these with diacritics on some letters?  They do not appear to be sorting in true alphabetical order when one is present.

"Individual bibs with more than one OCLC number" 
in KNXdb, sort by Title Brief, Deeds comes before Découverte:
![screen shot 2016-08-10 at 8 58 49 am](https://cloud.githubusercontent.com/assets/17598357/17556364/c43368ee-5ed8-11e6-9a16-ae7249d820ce.png)

"Title within a call number range at a specific location"
in KNXdb, Location=HSL Main Stacks, call number start=PA, call number end=PD, sort by Title Brief, three examples where they don't seem to be in true alpha order:
![screen shot 2016-08-10 at 9 06 00 am](https://cloud.githubusercontent.com/assets/17598357/17556767/2ed15a5c-5eda-11e6-8c64-099521cfcd81.png)
-o-o-o-o-o-o-o-o-
![screen shot 2016-08-10 at 9 06 37 am](https://cloud.githubusercontent.com/assets/17598357/17556771/341f246c-5eda-11e6-94ec-cd06b4bbc746.png)
-o-o-o-o-o-o-o-o-
![screen shot 2016-08-10 at 9 07 42 am](https://cloud.githubusercontent.com/assets/17598357/17556775/37f7d99e-5eda-11e6-8bc5-ac97030f4ae0.png)

**Comment by patrickzurek** [comment]
Using the last example you provided, could you tell me how you would expect it to have been sorted?  What's wrong with the way it is sorted?

**Comment by gibsonjc** [comment]
Just like the first one, there are three examples in titles in this KNXdb "Title within a call number range at a specific location" report that are not in alphabetical order:

Criticisms
Crito
Crítica (Críti comes incorrectly after Crito)
Crowells

Dialogues 
Dialogus 
Diálogo (Diálogo and Diálogos come incorrectly after Dialogu*)
Diálogos
Diccionario

Georgics
Gesta
Gesta
Géographie (Géo comes incorrectly after Ges)
Glosario

**Comment by bgant** [comment]
Searching Google for "Linguistic Sort" might help. Here is a document I found that might be useful:

https://docs.oracle.com/database/121/NLSPG/ch5lingsort.htm

**Comment by patrickzurek** [comment]
The examples given on that site (https://docs.oracle.com/database/121/NLSPG/ch5lingsort.htm#GUID-3C5D3D15-A891-4449-A946-9A5283D07561) seem to jive with the sorted examples in the github issue.

The comment before that responding to my question simply repeated again that they were not in alphabetical order.  I get that.  Please help me understand what order you would have expected the words to be ordered in then and what makes the way they're currently ordered wrong.  This will help me troubleshoot and find the right collating algorithm to use.

Regarding the example I specifically asked about: as I understand it, alphabetical ordering of French words consider the letter with the diacritic the same as the underlying letter for purposes of ordering so I don't understand what is wrong with the ordering in that example.

Thanks

**Comment by patrickzurek** [comment]
@gibsonjc  Sorry!  I just noticed in your last comment that you did include remarks in parenthesis indicating what was wrong with each example.  My mistake for not reading more closely.

**Comment by patrickzurek** [comment]
tl;dr:  I think the sorting is working as intended.  Characters with diacritical marks should alphabetically come after unadorned characters (except in certain languages where a distinction is made).  

I have not been able to find a truly authoritative source on the alphabetical ordering of words with diacritics, unless you consider wikipedia to be a definitive source.  If you can find something more authoritative, let me know.  This is what I've been able to find out:
- "Languages that treat accented letters as variants of the underlying letter usually alphabetize words with such symbols immediately after similar unmarked words." [1]
- > Letters modified by diacritical marks and ligatures of two letters should be arranged like their nearest basic equivalent letters in the English alphabet, for example:
  > 
  > These modified letters: 
  > á, à, â, a, å, ä, a
  > 
  > are arranged as: 
  > a  [2]
- "Letters with diacritical marks are to be sorted as the basis letter (the primary letter) without a diacritical mark (with the exceptions mentioned above for individual languages). But when two text strings are equal without regarding the diacritical marks, the diacritical marks are decisive for the sequence. _A letter without diacritical mark will always precede the same letter with a diacritical mark._" (emphasis mine) [3]

And some other sources (blog posts etc.) too that I forgot to record the source of.  More importantly, I don't recall encountering at all a single source where the ordering of words with diacritics was contrary to the sources listed above.

Finally, and again from wikipedia [3]:  

> Different languages use different rules to put diacritic characters in alphabetical order. French treats letters with diacritical marks the same as the underlying letter for purposes of ordering and dictionaries.
> 
> The Scandinavian languages, by contrast, treat the characters with diacritics ä, ö and å as new and separate letters of the alphabet, and sort them after z. Usually ä is sorted as equal to æ (ash) and ö is sorted as equal to ø (o-slash). Also, aa, when used as an alternative spelling to å, is sorted as such. Other letters modified by diacritics are treated as variants of the underlying letter, with the exception that ü is frequently sorted as y.
> 
> Languages that treat accented letters as variants of the underlying letter usually alphabetize words with such symbols immediately after similar unmarked words.

So, with specific regards to the four words with diacritical marks in the examples above (French, Portuguese, Portuguese, French, respectively, AFAIK): the French words are addressed specifically in the first line of the above quotation taken from wikipedia.  From [Portuguese orthography](https://en.wikipedia.org/wiki/Portuguese_orthography): "Accented letters and digraphs are not counted as separate characters for collation purposes."

Lost post short: I think the ordering is correct for the examples presented.  

Thoughts?

BTW I tried looking through the documentation/specification for the Unicode Collation Algorithm but I could not find anything definitive.  I'm sure something must be in there somewhere but I'm not smart enough to find it (or I skimmed it too quickly).  Also, regarding the UCA, we could switch our collating algorithm to that and use en_US as the locale but wouldn't we end up with the same results?  Isn't one of the points of using the UCA to be able to switch to the locale specific for the language we want to sort?  I think giving users the option to select the locale to sort report output is beyond the scope of this project.  I'm including this last paragraph in case someone more knowledgable than me about collating algorithms can correct me where I've undoubtably misunderstood something.

[1] https://en.wikipedia.org/wiki/Diacritic
[2] http://www.niso.org/publications/tr/tr03.pdf
[3] http://www.rostra.dk/alphabet/alpha_en.htm

**Comment by gibsonjc** [comment]
This may not be a huge problem in the grand scheme of things, but I was comparing the web-reports results to those received from MS Access as we have been doing for all reports.  Access sorts alphabetically without regard for diacritics, so it does make the results different if you are sorting by Title (or any other text with diacritics in it, I suppose).  (See screenshots from Access of the same three example results in KNXdb "Title within a call number range at a specific location" below.)

For Voyager reporting using MS Access, in order to get the diacritics in data to display at all, people have to adjust the character set in NLS_LANG in their PC's Registry to AMERICAN_AMERICA.US7ASCII.  In case you are interested, the instructions we have for doing this are at:
https://www.carli.illinois.edu/sites/files/i-share/documentation/secure/charsetoracle10g.pdf

![screen shot 2016-08-17 at 7 51 13 am](https://cloud.githubusercontent.com/assets/17598357/17738106/85609132-6455-11e6-8768-c22b96cc796f.png)

![screen shot 2016-08-17 at 7 51 45 am](https://cloud.githubusercontent.com/assets/17598357/17738111/8a5e1254-6455-11e6-9654-09fffe007be2.png)

![screen shot 2016-08-17 at 7 52 16 am](https://cloud.githubusercontent.com/assets/17598357/17738116/8e94fe00-6455-11e6-90b2-2ec03589665a.png)

**Comment by patrickzurek** [comment]
I agree that it's not a huge deal.  I pounced on the opportunity to research it further yesterday since I didn't have ssh access for a few hours.

So it looks like in the screenshot above that the characters with diacritics are being sorted before the   underlying letter instead of after?  Is that the only difference?  

If that's the case, is that the only difference between web-reports and Access output that you've encountered?  

To get web-reports output to truly match that in Access the only way forward I see is to either write our own collation algorithm (not even going to go there) or modify an existing one like https://github.com/jtauber/pyuca.  I don't know how hard that would be but I think it would be more work than we'd want to devote to this issue.  However I'll defer to you.  Should we just leave this issue alone and perhaps document the discrepancy on the web-reports documentation page?  

Otherwise, I can do some research into pyuca and see how hard it would be to modify it if you think getting web-reports to match Access is important enough to warrant it.

**Comment by gibsonjc** [comment]
On 8/15 we discussed in our meeting that we'd like for the sorting to be alphabetical as if there aren't any diacritics present (so that the diacritics aren't used for sorting at all).  PZ will look into normalization settings for the results data.

I will look for some examples of other unusual characters like æ to see how they are currently sorting.

**Comment by gibsonjc** [comment]
UIUdb bib 8201491 the title starts with æ.

The three "Bib Record without UT..." (Local Cat Maint.) queries return a lot of titles with diacritics for future testing.

**Comment by patrickzurek** [comment]
We discussed during today's meeting that the method of sorting employed on devel will be pushed to prod putting the servers into sync.

**Comment by patrickzurek** [comment]
I've moved the change from devel to prod.  Both servers should now be sorting fields identically.  

It took longer b/c I got sidetracked browsing some interesting projects that were using Django to connect to Oracle w/ US7ASCII encoding and then lunch....

There were a couple users on the system, but I did a handful of Apache restarts on devel while exercising different parts of web-reports and wasn't able to interrupt anything so I felt confident restarting Apache at will.  It looks like the Angular refactoring we did a while back ensure each page load has a permanent URL is working well in that regard.

Please leave this issue open until the next meeting.

**Comment by patrickzurek** [comment]
Sorry for the extra note, I want to be clear:  devel and prod are now in sync and web-reports on both servers should behave the same.
